comm/src/main/protobuf/coop/rchain/comm/protocol/routing.proto

```diff
@@ -46,24 +47,17 @@ message LookupResponse {
     repeated Node nodes = 1;
 }
 
-message Handshake {
-}
-
-message HandshakeResponse {
-}
-
 message Protocol {
     Header header                      = 1;
     ReturnHeader return_header         = 2;
 
     oneof message {
-        Hello             hello              = 3;
-        Disconnect        disconnect         = 4;
-        Ping              ping               = 5;
-        Pong              pong               = 6;
-        Lookup            lookup             = 7;
-        LookupResponse    lookup_response    = 8;
-        Handshake         handshake          = 9;
-        HandshakeResponse handshake_response = 10;
+        Hello               hello           = 3;
+        Disconnect          disconnect      = 4;
+        Ping                ping            = 5;
+        Pong                pong            = 6;
+        Lookup              lookup          = 7;
+        LookupResponse      lookup_response = 8;
+        google.protobuf.Any upstream        = 9;
```

Comment:
Can you say a little more for me about how chaining works?

Reply:
Things are set up with "levels" of protocol. At the bottom level, there is the mesh overlay, which is responsible for keeping last-seen times up-to-date and for responding to lookup queries. All the different messages that might come in are contained (effectively) in a union here.

In order to allow other protocols to be added on top of this, the schema must be kept separate. There is no facility like inheritance for protocol buffer schema, so an unspecified message type (`Any`, not Scala's `Any`, but protobuf's `Any`) can be used instead of one of those lower-level ping/pong-type messages. When a message of `Any` type is seen, if there is a `Dispatcher` type registered [here](https://github.com/rchain/rchain/pull/42/files#diff-2e4dabadc19c5d94097e2706f1cc0a19R14), then the message is passed up the chain [here](https://github.com/rchain/rchain/pull/42/files#diff-2e4dabadc19c5d94097e2706f1cc0a19R64).

It is really important that the rchain protocol level make decisions about whether some unknown node should be included in the mesh. But, if you want to use the rchain comm layer to build, say, a peer-to-peer file-sharing system, you want to include a different higher-level protocol (or no higher-level protocol at all).

Currently, this only matters for asynchronous messages, since responses to synchronous messages are delivered directly to the caller.

The cryptic note in the commit message about extending a trait refers to the existence of a field-level schema property that allows one to specify some information about the host language type of the message. That's lightly covered on [this page](https://scalapb.github.io/customizations.html). This might make things smoother in a statically typed language like Scala (and it also might not).